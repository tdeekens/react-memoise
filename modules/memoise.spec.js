import React from 'react';
import { mount } from 'enzyme';
import Memoise from './memoise';

describe('rendering', () => {
  describe('with FaaC', () => {
    let props;
    let wrapper;

    beforeEach(() => {
      const Component = props => (
        <Memoise args={[props.a, props.b]} compute={(a, b) => a * b}>
          {computedValue => <div>{computedValue}</div>}
        </Memoise>
      );
      wrapper = mount(<Component a={3} b={4} />);
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the computed value', () => {
      expect(wrapper).toHaveText('12');
    });
  });

  describe('with render-prop', () => {
    let props;
    let wrapper;

    beforeEach(() => {
      const Component = props => (
        <Memoise
          args={[props.a, props.b]}
          compute={(a, b) => a * b}
          render={computedValue => <div>{computedValue}</div>}
        />
      );
      wrapper = mount(<Component a={3} b={4} />);
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the computed value', () => {
      expect(wrapper).toHaveText('12');
    });
  });

  describe('with component', () => {
    let props;
    let wrapper;
    let ChildComponent;
    beforeEach(() => {
      ChildComponent = props => <div>{props.computedValue}</div>;
      const Component = props => (
        <Memoise
          args={[props.a, props.b]}
          compute={(a, b) => a * b}
          component={ChildComponent}
        />
      );
      wrapper = mount(<Component a={3} b={4} />);
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the computed value', () => {
      expect(wrapper).toHaveText('12');
    });
  });

  describe('when changing props', () => {
    let renderSpy;
    let computeSpy;
    let props;
    let wrapper;

    beforeEach(() => {
      renderSpy = jest.fn();
      computeSpy = jest.fn();
      const Component = props => (
        <Memoise
          args={[props.a, props.b]}
          compute={(a, b) => {
            computeSpy(a, b);
            return { value: a * b };
          }}
        >
          {memoised => {
            renderSpy(memoised);
            return <div>{memoised.value}</div>;
          }}
        </Memoise>
      );
      wrapper = mount(<Component a={3} b={4} />);
    });

    describe('to the same value', () => {
      beforeEach(() => {
        wrapper.setProps({ a: 3 });
      });

      it('should invoke render twice', () => {
        expect(renderSpy).toHaveBeenCalledTimes(2);
      });

      it('should invoke `compute` once', () => {
        expect(computeSpy).toHaveBeenCalledTimes(1);
      });
    });

    describe('to a different value', () => {
      beforeEach(() => {
        wrapper.setProps({ a: 7 });
      });

      it('should invoke render twice', () => {
        expect(renderSpy).toHaveBeenCalledTimes(2);
      });

      it('should invoke `compute` twice', () => {
        expect(computeSpy).toHaveBeenCalledTimes(2);
      });
    });
  });
});
