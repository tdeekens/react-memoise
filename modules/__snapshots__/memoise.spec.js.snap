// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering with FaaC should match snapshot 1`] = `
<Component
  a={3}
  b={4}
>
  <Memoise
    args={
      Array [
        3,
        4,
      ]
    }
    component={null}
    compute={[Function]}
    render={null}
  >
    <div>
      12
    </div>
  </Memoise>
</Component>
`;

exports[`rendering with component should match snapshot 1`] = `
<Component
  a={3}
  b={4}
>
  <Memoise
    args={
      Array [
        3,
        4,
      ]
    }
    component={[Function]}
    compute={[Function]}
    render={null}
  >
    <ChildComponent
      computedValue={12}
    >
      <div>
        12
      </div>
    </ChildComponent>
  </Memoise>
</Component>
`;

exports[`rendering with render-prop should match snapshot 1`] = `
<Component
  a={3}
  b={4}
>
  <Memoise
    args={
      Array [
        3,
        4,
      ]
    }
    component={null}
    compute={[Function]}
    render={[Function]}
  >
    <div>
      12
    </div>
  </Memoise>
</Component>
`;
